<a href="https://colab.research.google.com/github/adnaksbhat/Colab_JupyterNotebooks/blob/main/wow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Impacted Items Notifier using ECN

##Impacted notifier

In [ ]:
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df=pd.read_excel('/content/All_ECN.xlsx')

In [ ]:
df_ecr=pd.read_excel('/content/All_ECR.xlsx')

In [ ]:
god_cols=['ECR_CODE','COSTING_PROJECTS','CHANGE DESCRIPTION','IMPACTEDITEM','DIVISION']

In [ ]:
df_ecr=df_ecr[god_cols]

In [ ]:
df['COSTING_PROJECTS']=df['COSTING_PROJECTS'].astype(str)
df['ECN_DESC']=df['ECN_DESC'].astype(str)
df['IMPACTED_ITEMS']=df['IMPACTED_ITEMS'].astype(str)

df_ecr['COSTING_PROJECTS']=df_ecr['COSTING_PROJECTS'].astype(str)
df_ecr['CHANGE DESCRIPTION']=df_ecr['CHANGE DESCRIPTION'].astype(str)
df_ecr['IMPACTEDITEM']=df_ecr['IMPACTEDITEM'].astype(str)

In [ ]:
df_ecr['CHANGE DESCRIPTION'] = df_ecr['CHANGE DESCRIPTION'].str.replace('_x000D_', '')

In [ ]:
df_ecr['CHANGE DESCRIPTION'] = df_ecr['CHANGE DESCRIPTION'].str.replace('\n', ' ')

In [ ]:
df['ECN_DESC']=df['ECN_DESC'].str.replace('\n',' ')
df['ECN_DESC']=df['ECN_DESC'].str.replace('_x000D_', '')

In [ ]:
df_ecr.head()

ECR_CODE COSTING_PROJECTS  \
0  ECR-356880          1209501   
1  ECR-356877          5847426   
2  ECR-356868          1606025   
3  ECR-356867          1NG0216   
4  ECR-356862           506053   

                                  CHANGE DESCRIPTION             IMPACTEDITEM  \
0                    revised after customer comments           SOA1513365/001   
1                  Remove Item Eng in BOM SMO0398124    M67508006_5847426/003   
2       Flushing kit piping arrangement modification       BOM_SMO0417502/000   
3  Add manual valves under section "piping and va...  BOM_LCOIPHRS0241102/000   
4                         Revisione MR bolts/gaskets           SKO2103175/001   

            DIVISION  
0  OTHER ASSIGNMENTS  
1                UPG  
2                 GT  
3  OTHER ASSIGNMENTS  
4        PIPE-LAYOUT

In [ ]:
df.head()

ECN_CODE                           ECN_TYPE  \
0  RCID113996  Item Revision Change Notice(RCID)   
1  CIS-019844                Scope Change Notice   
2  CIS-019845                Scope Change Notice   
3  RCID113995  Item Revision Change Notice(RCID)   
4  RCID113997  Item Revision Change Notice(RCID)   

                                            ECN_DESC COMMENT COSTING_PROJECTS  \
0  Sow is defined and hrs are allocated. so doc i...     NaN          1106653   
1                                 APPROVED---> 1 CIS     NaN          5841560   
2                  update spare parts ad per 1209552     NaN          1209567   
3  Sow is defined and hrs are enabled for this jo...     NaN          1106654   
4  VDR Document needs to be updated with customer...     NaN          1107551   

  PROBLEM_ITEMS                               IMPACTED_ITEMS  \
0           NaN                               SOS4578300/000   
1           NaN  S34105970_5841560/000,S34105970_5841560/001   
2           NaN  P009A0001_1209567/000,P000BM999_1209567/000   
3           NaN                               SOS4578301/000   
4           NaN                               SOS4578332/000   

                                SOLUTION_ITEMS ECR OFFICE Name  
0                               SOS4578300/001          SYS-OP  
1  S34105970_5841560/003,S34105970_5841560/002             UPG  
2  P000BM999_1209567/001,P009A0001_1209567/001             PVS  
3                               SOS4578301/001          SYS-OP  
4                               SOS4578332/001          SYS-OP

In [ ]:
import pandas as pd

def merge_ecr_to_df(df, df_ecr):
    """
    Merges rows of df_ecr to df based on matching criteria.
    """

    # Create temporary merge keys in both DataFrames
    df['merge_key'] = df['COSTING_PROJECTS'].astype(str) + df['ECN_DESC'].astype(str)
    df_ecr['merge_key'] = df_ecr['COSTING_PROJECTS'].astype(str) + df_ecr['CHANGE DESCRIPTION'].astype(str)

    # Merge the DataFrames on the merge key
    merged_df = pd.merge(df, df_ecr, on='merge_key', how='left', suffixes=('', '_ecr'))

    # Drop the temporary merge key column
    merged_df.drop(columns=['merge_key'], inplace=True)

    return merged_df

# Example usage
merged_df = merge_ecr_to_df(df, df_ecr)


In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102646 entries, 0 to 102645
Data columns (total 14 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   ECN_CODE              102646 non-null  object
 1   ECN_TYPE              102646 non-null  object
 2   ECN_DESC              102646 non-null  object
 3   COMMENT               46 non-null      object
 4   COSTING_PROJECTS      102646 non-null  object
 5   PROBLEM_ITEMS         117 non-null     object
 6   IMPACTED_ITEMS        102646 non-null  object
 7   SOLUTION_ITEMS        100603 non-null  object
 8   ECR OFFICE Name       68871 non-null   object
 9   ECR_CODE              30248 non-null   object
 10  COSTING_PROJECTS_ecr  30248 non-null   object
 11  CHANGE DESCRIPTION    30248 non-null   object
 12  IMPACTEDITEM          30248 non-null   object
 13  DIVISION              30248 non-null   object
dtypes: object(14)
memory usage: 11.0+ MB


In [ ]:
merged_df['ECN_CODE']=merged_df['ECN_CODE'].astype(str)
merged_df['ECN_TYPE']=merged_df['ECN_TYPE'].astype(str)
merged_df['ECN_DESC']=merged_df['ECN_DESC'].astype(str)
merged_df['COMMENT']=merged_df['COMMENT'].astype(str)
merged_df['COSTING_PROJECTS']=merged_df['COSTING_PROJECTS'].astype(str)
merged_df['PROBLEM_ITEMS']=merged_df['PROBLEM_ITEMS'].astype(str)
merged_df['IMPACTED_ITEMS']=merged_df['IMPACTED_ITEMS'].astype(str)
merged_df['SOLUTION_ITEMS']=merged_df['SOLUTION_ITEMS'].astype(str)

merged_df['ECR OFFICE Name']=merged_df['ECR OFFICE Name'].astype(str)
merged_df['ECR_CODE']=merged_df['ECR_CODE'].astype(str)
merged_df['COSTING_PROJECTS_ecr']=merged_df['COSTING_PROJECTS_ecr'].astype(str)
merged_df['CHANGE DESCRIPTION']=merged_df['CHANGE DESCRIPTION'].astype(str)
merged_df['IMPACTEDITEM']=merged_df['IMPACTEDITEM'].astype(str)
merged_df['DIVISION']=merged_df['DIVISION'].astype(str)



In [ ]:
merged_df["CHANGE DESCRIPTION"] = merged_df["CHANGE DESCRIPTION"].str.replace('\n', ' ')

In [ ]:
merged_df.head()

ECN_CODE                           ECN_TYPE  \
0  RCID113996  Item Revision Change Notice(RCID)   
1  CIS-019844                Scope Change Notice   
2  CIS-019845                Scope Change Notice   
3  RCID113995  Item Revision Change Notice(RCID)   
4  RCID113997  Item Revision Change Notice(RCID)   

                                            ECN_DESC COMMENT COSTING_PROJECTS  \
0  Sow is defined and hrs are allocated. so doc i...     nan          1106653   
1                                 APPROVED---> 1 CIS     nan          5841560   
2                  update spare parts ad per 1209552     nan          1209567   
3  Sow is defined and hrs are enabled for this jo...     nan          1106654   
4  VDR Document needs to be updated with customer...     nan          1107551   

  PROBLEM_ITEMS                               IMPACTED_ITEMS  \
0           nan                               SOS4578300/000   
1           nan  S34105970_5841560/000,S34105970_5841560/001   
2           nan  P009A0001_1209567/000,P000BM999_1209567/000   
3           nan                               SOS4578301/000   
4           nan                               SOS4578332/000   

                                SOLUTION_ITEMS ECR OFFICE Name    ECR_CODE  \
0                               SOS4578300/001          SYS-OP         nan   
1  S34105970_5841560/003,S34105970_5841560/002             UPG  ECR-356849   
2  P000BM999_1209567/001,P009A0001_1209567/001             PVS         nan   
3                               SOS4578301/001          SYS-OP         nan   
4                               SOS4578332/001          SYS-OP         nan   

  COSTING_PROJECTS_ecr  CHANGE DESCRIPTION           IMPACTEDITEM DIVISION  
0                  nan                 nan                    nan      nan  
1              5841560  APPROVED---> 1 CIS  S34105970_5841560/001      UPG  
2                  nan                 nan                    nan      nan  
3                  nan                 nan                    nan      nan  
4                  nan                 nan                    nan      nan

In [ ]:
def find_missing_impacts(row):

  impacted_items = str(row['IMPACTED_ITEMS']).split(',')  # Split comma-separated values
  impacted_item = str(row['IMPACTEDITEM'])
  missing = [item for item in impacted_items if item != impacted_item and item != 'nan']
  # Find items not in 'IMPACTEDITEM' and not 'nan'

  return ','.join(missing)  # Join missing impacts with commas





merged_df['missing_impacts'] = merged_df.apply(find_missing_impacts, axis=1)


In [ ]:
merged_df[merged_df['ECN_CODE']=='CIS-019854']

ECN_CODE             ECN_TYPE  \
7853  CIS-019854  Scope Change Notice   

                                               ECN_DESC COMMENT  \
7853  Aggiungere il dummy del bilancino M11101534 so...     nan   

     COSTING_PROJECTS PROBLEM_ITEMS         IMPACTED_ITEMS  \
7853          1704047           nan  M111AE999_1604379/002   

             SOLUTION_ITEMS ECR OFFICE Name    ECR_CODE COSTING_PROJECTS_ecr  \
7853  M111AE999_1604379/003             nan  ECR-356851              1704047   

                                     CHANGE DESCRIPTION  \
7853  Aggiungere il dummy del bilancino M11101534 so...   

               IMPACTEDITEM DIVISION        missing_impacts  
7853  M111AE999_1704047/012     CORE  M111AE999_1604379/002

In [ ]:
print(merged_df[merged_df['ECN_CODE']=='CIS-019854'].to_string())

        ECN_CODE             ECN_TYPE                                                                                            ECN_DESC COMMENT COSTING_PROJECTS PROBLEM_ITEMS         IMPACTED_ITEMS         SOLUTION_ITEMS ECR OFFICE Name    ECR_CODE COSTING_PROJECTS_ecr                                                                                  CHANGE DESCRIPTION           IMPACTEDITEM DIVISION        missing_impacts
7853  CIS-019854  Scope Change Notice  Aggiungere il dummy del bilancino M11101534 sotto il BOP Special Tools M111AE999 Ref. WF#202827089     nan          1704047           nan  M111AE999_1604379/002  M111AE999_1604379/003             nan  ECR-356851              1704047  Aggiungere il dummy del bilancino M11101534 sotto il BOP Special Tools M111AE999 Ref. WF#202827089  M111AE999_1704047/012     CORE  M111AE999_1604379/002


In [ ]:
print(df_ecr[df_ecr['ECR_CODE']=='ECR-356732'].to_string())

      ECR_CODE COSTING_PROJECTS                                         CHANGE DESCRIPTION   IMPACTEDITEM           DIVISION                                                         merge_key
29  ECR-356732          1NG0209  ajout: banc de maintenance bunle caisse à outils équippée  CLMST0244/000  OTHER ASSIGNMENTS  1NG0209ajout: banc de maintenance bunle caisse à outils équippée


In [ ]:
merged_df['COSTING_PROJECTS'] = merged_df['COSTING_PROJECTS'].astype(str)

In [ ]:
!pip install fuzzywuzzy[speedup]

##search method

In [ ]:
from __future__ import division
import ipywidgets as widgets
from IPython.display import display

import pandas as pd
import numpy as np
import re
import unicodedata
from fuzzywuzzy import fuzz

# Create input widgets
costing_projects_input = widgets.Text(description="JobNo:")
change_description_input = widgets.Text(description="Description:")
impacted_items_input = widgets.Text(description="Impacted Items:")
#division_input = widgets.Text(description="Division:")

search_button = widgets.Button(description="Search")

# Create output widget
output_widget = widgets.Output()



def normalize_text(text):
    text = unicodedata.normalize("NFKD", text)  # Normalize to remove special characters
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    text = text.strip()  # Remove leading/trailing spaces
    return text

merged_df["CHANGE DESCRIPTION"] = merged_df["CHANGE DESCRIPTION"].astype(str).apply(normalize_text)
merged_df["IMPACTED_ITEMS"] = merged_df["IMPACTED_ITEMS"].astype(str).apply(normalize_text)
merged_df["COSTING_PROJECTS"] = merged_df["COSTING_PROJECTS"].astype(str).apply(normalize_text)


# Define search function
def search_df(button):
  with output_widget:
    output_widget.clear_output()
    costing_projects = str(costing_projects_input.value)
    change_description = normalize_text(change_description_input.value).replace('\n', '')
    #division=division_input.value
    impacted_items = str(impacted_items_input.value)






    # Filter the DataFrame
    pattern = '|'.join(impacted_items.split(','))  # Create regex pattern for any of the impacted items
    filtered_df = merged_df[
        (merged_df["COSTING_PROJECTS"] == costing_projects) &
        #(merged_df["CHANGE DESCRIPTION"] == change_description) &
        (merged_df["CHANGE DESCRIPTION"].apply(lambda x: fuzz.ratio(normalize_text(x), change_description) >= 90)) &
        #(merged_df["DIVISION"]==division)&
        (merged_df["IMPACTED_ITEMS"].str.contains(pattern, regex=True))
    ]



    if not filtered_df.empty:
          impacted_items_list = impacted_items.split(',')  # Split comma-separated values
          all_impacted_items = []
          for index in filtered_df.index:
            all_impacted_items.extend(str(filtered_df.loc[index, 'IMPACTED_ITEMS']).split(','))  # Get all impacted items from filtered rows

          # Find items not in impacted_items_list and not 'nan'
          remaining_missing = [item for item in all_impacted_items if item not in impacted_items_list and item != 'nan']
          remaining_missing = list(set(remaining_missing))  # Remove duplicates

          print(f"Remaining Missing Impacts: {','.join(remaining_missing)}")  # Join missing impacts with commas
    else:
          print("No matching records found.")
# search button

search_button.on_click(search_df)

# Display widgets


##output

In [ ]:
merged_df[(merged_df['COSTING_PROJECTS']=='1NG0218') & (merged_df['IMPACTEDITEM']=='ICSSH0031102/001')]

ECN_CODE                           ECN_TYPE  \
6649  RCID112651  Item Revision Change Notice(RCID)   

                                               ECN_DESC COMMENT  \
6649  Filter element material: stainless steel 316. ...     nan   

     COSTING_PROJECTS PROBLEM_ITEMS  \
6649          1NG0218           nan   

                                         IMPACTED_ITEMS  \
6649  ICSSH0031102/001,ICSSH1011102/001,ICSSH2011102...   

                                         SOLUTION_ITEMS ECR OFFICE Name  \
6649  ICSSH2011102/002,ICSSH1011102/002,ICSSH1021102...             ASE   

        ECR_CODE COSTING_PROJECTS_ecr  \
6649  ECR-352610              1NG0218   

                                     CHANGE DESCRIPTION      IMPACTEDITEM  \
6649  Filter element material: stainless steel 316. ...  ICSSH0031102/001   

               DIVISION                                    missing_impacts  
6649  OTHER ASSIGNMENTS  ICSSH1011102/001,ICSSH2011102/001,ICSSH1021102...

In [ ]:
display(costing_projects_input, change_description_input, impacted_items_input, search_button, output_widget)

Text(value='', description='JobNo:')

Text(value='', description='Description:')

Text(value='', description='Impacted Items:')

Button(description='Search', style=ButtonStyle())

Output()

In [ ]:
'''
1606624

1. Gas detector rack dimensions are updated in order to avoid interference\n2. Flowmeter VDR updated fro customer comments


1704047

Aggiungere il dummy del bilancino M11101534 sotto il BOP Special Tools M111AE999Ref. WF#202827089



5849119
due to Main motor shaft modification, we need to make the coupling guard shorter (20 mm); to avoid issue on casting body, we prefer to modify the axial dilatation joint

1604379
Change required to update P&ID with Customer comments
SOM6906969/001



1NG0218
Filter element material: stainless steel 316.
Minimum temperature changed, already discussed with supplier no material impact.
ICSSH0031102/001

'''

In [ ]:
from google.colab import files

df.to_excel('ECN_ECR_wow.xlsx', index=False)

files.download('ECN_ECR_wow.xlsx')

# Discipline | Activity

##data processing

In [1]:
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df=pd.read_excel('/content/All_ECR.xlsx')

In [2]:
df.head()

ECR_CODE                    ECR_TYPE  \
0  ECR-356880  Requisition Change Request   
1  ECR-356877  Requisition Change Request   
2  ECR-356868  Requisition Change Request   
3  ECR-356867  Requisition Change Request   
4  ECR-356862  Requisition Change Request   

                               SOURCE OF CHANGE                   REASON_DESC  \
0                  CHANGE DRIVEN BY ENGINEERING    DESIGN PROGRESS COMPLETION   
1                  CHANGE DRIVEN BY ENGINEERING                  TYPO MISTAKE   
2  CHANGE DRIVEN BY PROJECT MANAGEMENT/PLANNING  FUNCTIONAL UNIT REALLOCATION   
3                  CHANGE DRIVEN BY ENGINEERING     PREDECESSOR DATA REVISION   
4                  CHANGE DRIVEN BY ENGINEERING    DESIGN PROGRESS COMPLETION   

   COMMENT  ACKNOWLEDGE_DATE  PROMISE_DATE     SSO ID ECR_CREATION_DATE  \
0      NaN               NaN           NaN  105053425        2019-12-30   
1      NaN               NaN           NaN  212362330        2019-12-30   
2      NaN               NaN           NaN  212334942        2019-12-30   
3      NaN               NaN           NaN  105059509        2019-12-30   
4      NaN               NaN           NaN  105704753        2019-12-30   

     CHANGE_TYPE ENG IMPACTED LIST COSTING_PROJECTS  \
0     FFF&P ONLY               NaN          1209501   
1  CIS/CIS+FFF&P               NaN          5847426   
2     FFF&P ONLY               NaN          1606025   
3     FFF&P ONLY               NaN          1NG0216   
4     FFF&P ONLY               NaN           506053   

                MATERIAL_CATEGORY   CHANGE_CATEGORY CHANGE_ON_RECEIVED_ITEMS  \
0  EQUIPMENT/MATERIAL FULL SUPPLY          DOCUMENT                       No   
1                           OTHER  REMOVE COMPONENT                       No   
2                   SPECIAL TOOLS        CHANGE QTY                       No   
3                          VALVES     ADD COMPONENT                       No   
4              PIPINGS & SUPPORTS        CHANGE QTY                       No   

  IMPACT VERIFICATION                                 CHANGE DESCRIPTION  \
0                 Yes                    revised after customer comments   
1                 Yes                  Remove Item Eng in BOM SMO0398124   
2                 Yes       Flushing kit piping arrangement modification   
3                 Yes  Add manual valves under section "piping and va...   
4                 Yes                         Revisione MR bolts/gaskets   

              IMPACTEDITEM                  IMPACTED ITEM DESCRIPTION  \
0           SOA1513365/001            SKID INSTRUMENTATION LAYOUT DWG   
1    M67508006_5847426/003                BOP - GT READY FOR SHIPMENT   
2       BOM_SMO0417502/000         MS5001 TURBOGENERATOR FLUSHING KIT   
3  BOM_LCOIPHRS0241102/000        PIPING AND VALVES FOR RUN DOWN TANK   
4           SKO2103175/001  MATERIAL REQUISITION G0201 BOLTS AND NUTS   

   FUNCTION           DIVISION  ECR CREATION YEAR  AGING   APPROVER  \
0  IVPG-ENG  OTHER ASSIGNMENTS               2019    1.5  105053425   
1   GEP-ENG                UPG               2019    1.5  212362330   
2   GEP-ENG                 GT               2019    1.5  105052434   
3    TD-ENG  OTHER ASSIGNMENTS               2019    1.5  105030743   
4   GEP-ENG        PIPE-LAYOUT               2019    1.5  105702671   

  APPROVER_DECISION APPROVER DESC DATE CLOSURE DISPOSITION  MATURITY  \
0           APPROVE         2019-12-30  Closed    Approved  Complete   
1           APPROVE         2019-12-30  Closed    Approved  Complete   
2           APPROVE         2019-12-31  Closed    Approved  Complete   
3           APPROVE         2019-12-30  Closed    Approved  Complete   
4           APPROVE         2019-12-30  Closed    Approved  Complete   

  AS_BUILT_FROM_SITE REFERENCE ECR INTERNAL_REFERENT EXCLUDED  \
0                 NO           NaN               NaN        N   
1                 NO           NaN               NaN        N   
2                 NO           Na

In [3]:
good_cols=['ECR_CODE','CHANGE_TYPE','COSTING_PROJECTS','IMPACTEDITEM','IMPACTED ITEM DESCRIPTION','DIVISION','ENG IMPACTED LIST']

In [4]:
df=df[good_cols]

In [5]:
df['IMPACTEDITEM_new']=df['IMPACTEDITEM']

<ipython-input-5-2ee7de3ae136>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['IMPACTEDITEM_new']=df['IMPACTEDITEM']


In [6]:
df["IMPACTEDITEM"] = df["IMPACTEDITEM"].astype(str)
df["IMPACTEDITEM_new"] = df["IMPACTEDITEM_new"].astype(str)

<ipython-input-6-39cf13568135>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["IMPACTEDITEM"] = df["IMPACTEDITEM"].astype(str)
<ipython-input-6-39cf13568135>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["IMPACTEDITEM_new"] = df["IMPACTEDITEM_new"].astype(str)


In [7]:
df['IMPACTEDITEM_new']=(df['IMPACTEDITEM_new'].str.split("/").str[0].str.replace('BOM_', ''))
df['IMPACTEDITEM_new']=(df['IMPACTEDITEM_new'].str.split("-").str[0])

In [8]:
df.head(2)

ECR_CODE    CHANGE_TYPE COSTING_PROJECTS           IMPACTEDITEM  \
0  ECR-356880     FFF&P ONLY          1209501         SOA1513365/001   
1  ECR-356877  CIS/CIS+FFF&P          5847426  M67508006_5847426/003   

         IMPACTED ITEM DESCRIPTION           DIVISION ENG IMPACTED LIST  \
0  SKID INSTRUMENTATION LAYOUT DWG  OTHER ASSIGNMENTS               NaN   
1      BOP - GT READY FOR SHIPMENT                UPG               NaN   

    IMPACTEDITEM_new  
0         SOA1513365  
1  M67508006_5847426

##ISPO data

In [9]:
import pandas as pd

df_all_ispo=pd.read_excel("/content/All_ISPO.xlsx")

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df_all_ispo.head(2)

PROJECT_ID PRIMARY_RESOURCE_NAME   ACT_ID                  ACTIVITY_NAME  \
0    1102467                   NaN  PC00S30  CECO DESIGN REVIEW 1 - REPORT   
1    1102467                   NaN  PC00S30  CECO DESIGN REVIEW 1 - REPORT   

  STEP_ID STEP_NAME IMPACTED_NCR EARLY_DATE LATE_FINISH PROMISE_DATE COMMENT  \
0     NaN       NaN          NaN 2009-05-04         NaT          NaT     NaN   
1     NaN       NaN          NaN 2009-05-04         NaT          NaT     NaN   

   ACKNOWLEDGE LAST MODIFIED DATE               PROJECT_NAME    ROLE  \
0          NaN                NaT  1102467-ADCO SAS GAS LIFT  CCCALC   
1          NaN                NaT  1102467-ADCO SAS GAS LIFT  CCCALC   

  PRIMARY_RESOURCE_SSO LONG_DUMMY  ACT_TYPE   REAL_CODE   EARLY_STATUS  DELAY  \
0                  NaN        NaN     510.0  SOS8629724  COMPLETEDLATE    NaN   
1                  NaN        NaN     510.0         NaN  COMPLETEDLATE    NaN   

  ACTUAL_FINISH      TYPE EARLY_DUE_WEEK AGING ITEM_AGGREGATION  \
0    2009-05-13  Activity             18   NaN              NaN   
1    2009-05-13      Step             18   NaN              NaN   

  P6-PROJ_CONTR_DELIVERY_DATE   PL  PO_NUMBER  PROJECT_MANAGER  \
0                         NaN  NaN        NaN              NaN   
1                         NaN  NaN        NaN              NaN   

   Completed Late               RAC_DT PLANNING_DATE TRAIN_TYPE SUB_PL  \
0             NaN  2009-02-17 00:00:00           NaT        NaN    NaN   
1             NaN  2009-02-17 00:00:00           NaT        NaN    NaN   

  ORGANIZATION_NAME SYSTEM_DATE PROJECT_PHASE CONTRACT_NUMBER JOB_TYPE  \
0               NaN         NaT        CLOSED       M_1102467      NaN   
1               NaN         NaT        CLOSED       M_1102467      NaN   

  DIVISION CONC  Backlog  EXCLUDE FUNCTION  PROJECT_ENGINEER  PROJECT_PLANNER  \
0      NaN  NaN      NaN      NaN      NaN               NaN              NaN   
1      NaN  NaN      NaN      NaN      NaN               NaN              NaN   

  DOC_TYPE LATE_STATUS MAIN_FUNCTION  Open PLANNING_FW UNIFIER_PROJECT_STATUS  \
0      NaN         NaN           NaN   NaN         NaN                    NaN   
1      NaN         NaN           NaN   NaN         NaN                    NaN   

   Completed_On_Time  SUM PO_SPECIFICATION_ID VENDOR SIGN-OFF_STATUS  \
0                NaN  NaN                 NaN    NaN             NaN   
1                NaN  NaN                 NaN    NaN             NaN   

  APPROVE_TYPE DOCUMENT_DUE_DATE  DWG_ISSUE_NUMBER_FINAL Design_Configurator  \
0          NaN               NaT                     NaN                 NaN   
1          NaN               NaT                     NaN                 NaN   

            PM  RETURN_DATE_PLANNED  BL_Project_Finish  RETURN_CODE  \
0  Tariq Ahmed                  NaN                NaN          NaN   
1  Tariq Ahmed                  NaN                NaN          NaN   

   DOC_RETURN_CODE  ACT_REGIONAL_ERP  ADDITIONAL_CUSTOMER_DOC  \
0              NaN               NaN                      NaN   
1              NaN               NaN                      NaN   

                      PE  ACT_AGGREGATION  GM  RETURN_DATE_ACTUAL  \
0  Laura Fernanda Grayeb              NaN NaN                 NaN   
1  Laura Fernanda Grayeb              NaN NaN                 NaN   

  PROJECT_STATUS_Primavera  FIRST_SUBMISSION  LAST_SUBMISSION  DELAY_CUSTOMER  \
0                      NaN               NaN              NaN             NaN   
1                      NaN               NaN              NaN             NaN   

   CUSTOMER_DOC  UNIFIER_DOCUMENT_STATUS  UNIFIER_REVISION_STATUS  \
0           NaN                      NaN                      NaN   
1           NaN                      NaN                      NaN   

   PLANNING_YEAR PLANNING_MONTH PLANNING QTR TEAM LEADER  ACT_DEPARTMENT  \
0            NaN            NaN          NaN         NaN             NaN   
1            NaN            NaN          NaN      

In [10]:
df_ispo=df_all_ispo[["PROJECT_ID","ACT_ID","ACTIVITY_NAME","LONG_DUMMY","REAL_CODE"]]

In [11]:
df_ispo.shape

(709450, 5)

In [12]:
# prompt: if same 'PROJECT_ID' and 'ACT_ID' are repeated in another row then drop the rows with REAL_CODE as NaN

import pandas as pd

# Assuming your DataFrame is named 'df'
def drop_nan_duplicates(df):
    """Drops rows with NaN REAL_CODE where PROJECT_ID and ACT_ID are duplicated."""

    # Group by 'PROJECT_ID' and 'ACT_ID'
    grouped = df.groupby(['PROJECT_ID', 'ACT_ID'])

    # Iterate through each group
    rows_to_drop = []
    for (project_id, act_id), group in grouped:
        if len(group) > 1:  # Check for duplicates
          # Identify rows with NaN REAL_CODE within the duplicate group
            nan_real_code_rows = group[group['REAL_CODE'].isna()]
            rows_to_drop.extend(nan_real_code_rows.index.tolist())

    # Drop identified rows from original dataframe
    df_cleaned = df.drop(index=rows_to_drop)
    return df_cleaned


df_ispo = drop_nan_duplicates(df_ispo)


In [13]:
df_ispo.shape

(627018, 5)

In [14]:
df_ispo.head()

PROJECT_ID   ACT_ID                  ACTIVITY_NAME LONG_DUMMY   REAL_CODE
0     1102467  PC00S30  CECO DESIGN REVIEW 1 - REPORT        NaN  SOS8629724
2     1102467  PC00U50            BASIC DESIGN REVIEW        NaN  SOK6747811
4     1102467  SP00S21      ROTATION SENSE DEFINITION        NaN  SOS9989279
6     1102467  DS00S42           FUNCTIONAL DATA FORM        NaN  SOK7352503
10    1102467  DS00S55         CC INTERNAL DATA SHEET        NaN  SOS9989297

In [15]:
# prompt: in df2 if 'REAL_CODE' is null or has 'DELETED IN TCE' then replace it with its corresponding 'LONG_DUMMY' value. if LONG_DUMMY value is also null then fill both 'REAL_CODE' and "LONG_DUMMY' as 'unknown'

# Assuming your DataFrame is named 'df2'
def fill_real_code(df2):
    for index, row in df2.iterrows():
        if pd.isnull(row['REAL_CODE']):
            if pd.notnull(row['LONG_DUMMY']) and row['LONG_DUMMY'] != '':
                df2.loc[index, 'REAL_CODE'] = row['LONG_DUMMY']
            else:
                df2.loc[index, 'REAL_CODE'] = 'unknown'
                df2.loc[index, 'LONG_DUMMY'] = 'unknown'
    return df2

df_ispo_2 = fill_real_code(df_ispo)

## merge ECR and ISPO

In [16]:
df_final = pd.merge(df, df_ispo_2,  left_on=['IMPACTEDITEM_new'], right_on=['REAL_CODE'], how='left')




In [17]:
from google.colab import files
df_final.to_csv('df_final_dis.csv')
files.download('df_final_dis.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [18]:
df_final.head()

ECR_CODE    CHANGE_TYPE COSTING_PROJECTS             IMPACTEDITEM  \
0  ECR-356880     FFF&P ONLY          1209501           SOA1513365/001   
1  ECR-356877  CIS/CIS+FFF&P          5847426    M67508006_5847426/003   
2  ECR-356868     FFF&P ONLY          1606025       BOM_SMO0417502/000   
3  ECR-356867     FFF&P ONLY          1NG0216  BOM_LCOIPHRS0241102/000   
4  ECR-356862     FFF&P ONLY           506053           SKO2103175/001   

                   IMPACTED ITEM DESCRIPTION           DIVISION  \
0            SKID INSTRUMENTATION LAYOUT DWG  OTHER ASSIGNMENTS   
1                BOP - GT READY FOR SHIPMENT                UPG   
2         MS5001 TURBOGENERATOR FLUSHING KIT                 GT   
3        PIPING AND VALVES FOR RUN DOWN TANK  OTHER ASSIGNMENTS   
4  MATERIAL REQUISITION G0201 BOLTS AND NUTS        PIPE-LAYOUT   

  ENG IMPACTED LIST   IMPACTEDITEM_new PROJECT_ID     ACT_ID  \
0               NaN         SOA1513365        NaN        NaN   
1               NaN  M67508006_5847426        NaN        NaN   
2               NaN         SMO0417502    1606025  M11101941   
3               NaN    LCOIPHRS0241102    1NG0214  S61000640   
4               NaN         SKO2103175     506053  K005PTA31   

                                   ACTIVITY_NAME            LONG_DUMMY  \
0                                            NaN                   NaN   
1                                            NaN                   NaN   
2          ENG-FLUSHING KIT FOR MINERAL LUBE OIL  M11101941010_1606025   
3        ENG PIPING AND VALVES FOR RUN DOWN TANK  S61000640010_1NG0214   
4  ENG - BOLTS AND BOLTS AND NUTS. - BALANCE MTO  K005PTA31010_0506053   

         REAL_CODE  
0              NaN  
1              NaN  
2       SMO0417502  
3  LCOIPHRS0241102  
4       SKO2103175

In [19]:
df_final.shape

(47089, 13)

## group DIVISION wise and IMAPCTEDFILE wise and DISCIPLINE wise

In [20]:

df_final['ACT_DIVISION'] = df_final.groupby('ACT_ID')['DIVISION'].transform(lambda x: ','.join(x.astype(str).unique()))

In [21]:
df_final['ACT_IMPACTED']=df_final.groupby('ACT_ID')['IMPACTEDITEM'].transform(lambda x: ','.join(x.astype(str).unique()))

In [22]:
df_final['ACT_DISCIPLINE']=df_final.groupby('ACT_ID')['ENG IMPACTED LIST'].transform(lambda x: ','.join(x.astype(str).unique()))

In [55]:
df_final.tail(2)

ECR_CODE CHANGE_TYPE COSTING_PROJECTS        IMPACTEDITEM  \
47087  ECR-399731  FFF&P ONLY          6240142        SR-54464/000   
47088  ECR-399729  FFF&P ONLY          8337156  BOM_NLA7158009/001   

             IMPACTED ITEM DESCRIPTION           DIVISION ENG IMPACTED LIST  \
47087  testata lavorazioni preliminari              RCRQS               NaN   
47088             SKID INSTRUMENTATION  OTHER ASSIGNMENTS               NaN   

      IMPACTEDITEM_new PROJECT_ID ACT_ID ACTIVITY_NAME LONG_DUMMY REAL_CODE  \
47087               SR        NaN    NaN           NaN        NaN       NaN   
47088       NLA7158009        NaN    NaN           NaN        NaN       NaN   

      ACT_DIVISION ACT_IMPACTED ACT_DISCIPLINE  
47087          NaN          NaN                 
47088          NaN          NaN

In [24]:
import pandas as pd

def get_unique_disciplines(df):
    """
    Extracts unique discipline values from the 'ACT_DISCIPLINE' column,
    handling potential comma-separated values within cells.
    """
    all_disciplines = []
    for value in df['ACT_DISCIPLINE']:
        if pd.isna(value): # handle NaN
            continue
        disciplines = str(value).split(',')
        for disc in disciplines:
            cleaned_disc = disc.strip()
            if cleaned_disc: #avoid empty string
                all_disciplines.append(cleaned_disc)

    unique_disciplines = list(set(all_disciplines))
    return unique_disciplines


unique_disciplines = get_unique_disciplines(df_final)




In [25]:
df_final['ACT_DISCIPLINE'] = df_final['ACT_DISCIPLINE'].apply(lambda x: ','.join(get_unique_disciplines(pd.DataFrame({'ACT_DISCIPLINE': [x]}))))

In [26]:
df_final.head()

ECR_CODE    CHANGE_TYPE COSTING_PROJECTS             IMPACTEDITEM  \
0  ECR-356880     FFF&P ONLY          1209501           SOA1513365/001   
1  ECR-356877  CIS/CIS+FFF&P          5847426    M67508006_5847426/003   
2  ECR-356868     FFF&P ONLY          1606025       BOM_SMO0417502/000   
3  ECR-356867     FFF&P ONLY          1NG0216  BOM_LCOIPHRS0241102/000   
4  ECR-356862     FFF&P ONLY           506053           SKO2103175/001   

                   IMPACTED ITEM DESCRIPTION           DIVISION  \
0            SKID INSTRUMENTATION LAYOUT DWG  OTHER ASSIGNMENTS   
1                BOP - GT READY FOR SHIPMENT                UPG   
2         MS5001 TURBOGENERATOR FLUSHING KIT                 GT   
3        PIPING AND VALVES FOR RUN DOWN TANK  OTHER ASSIGNMENTS   
4  MATERIAL REQUISITION G0201 BOLTS AND NUTS        PIPE-LAYOUT   

  ENG IMPACTED LIST   IMPACTEDITEM_new PROJECT_ID     ACT_ID  \
0               NaN         SOA1513365        NaN        NaN   
1               NaN  M67508006_5847426        NaN        NaN   
2               NaN         SMO0417502    1606025  M11101941   
3               NaN    LCOIPHRS0241102    1NG0214  S61000640   
4               NaN         SKO2103175     506053  K005PTA31   

                                   ACTIVITY_NAME            LONG_DUMMY  \
0                                            NaN                   NaN   
1                                            NaN                   NaN   
2          ENG-FLUSHING KIT FOR MINERAL LUBE OIL  M11101941010_1606025   
3        ENG PIPING AND VALVES FOR RUN DOWN TANK  S61000640010_1NG0214   
4  ENG - BOLTS AND BOLTS AND NUTS. - BALANCE MTO  K005PTA31010_0506053   

         REAL_CODE           ACT_DIVISION  \
0              NaN                    NaN   
1              NaN                    NaN   
2       SMO0417502                     GT   
3  LCOIPHRS0241102  OTHER ASSIGNMENTS,CST   
4       SKO2103175            PIPE-LAYOUT   

                                        ACT_IMPACTED ACT_DISCIPLINE  
0                                                NaN                 
1                                                NaN                 
2  BOM_SMO0417502/000,BOM_SSO2493917/003,BOM_SSO2...            nan  
3  BOM_LCOIPHRS0241102/000,SKO0061411/000,BOM_SKO...            nan  
4       SKO2103175/001,SKO2103174/000,SKO2103175/003            nan

In [27]:
from google.colab import files
df_final.to_csv('Disciplines_impacted.csv')
files.download('Disciplines_impacted.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## take only good cols and not null values

Total ECR rows -> **47089**


No. of rows with null for :

ActIds is :  **29749**

Eng Impacted List is : **41610**


#==================#
After grouping Disciplines and Impacted Items :
__________________
No . unique Activity ID  rows : **3460**

No. of rows with Disciplines as null : **2615**

So, total non null rows -> 845

__________________________




In [37]:
df_final['ACT_ID'].isnull().sum()


29749

In [38]:
df_final['ENG IMPACTED LIST'].isnull().sum()

41610

In [56]:
new_cols=['ACT_ID','ACTIVITY_NAME','ACT_DIVISION','ACT_IMPACTED','ACT_DISCIPLINE']

In [57]:
dfdf=df_final[new_cols]

In [63]:
dfdf.shape

(3460, 5)

In [60]:
dfdf.drop_duplicates(inplace=True)

<ipython-input-60-9803d1c14469>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfdf.drop_duplicates(inplace=True)


In [62]:
dfdf=dfdf[dfdf['ACT_ID'].isna()==False]

In [64]:
dfdf.tail(2)

ACT_ID                       ACTIVITY_NAME ACT_DIVISION  \
47052  M64008610  ELECTRICAL PACKAGING SPECIFICATION          UPG   
47081  N34A06584            ENG-MISCELLANEA FOR RING         TECH   

                                ACT_IMPACTED ACT_DISCIPLINE  
47052  BOM_SMO0126889/000,BOM_SMO0126629/000            nan  
47081                     BOM_SNO3704951/000            nan

After grouping :

In [35]:
from google.colab import files
dfdf.to_csv('Disciplines_impacted_notnull.csv')
files.download('Disciplines_impacted_notnull.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>